### PR TITLE
INS-3832: pulse fetching from light or heavy for request

### PR DESCRIPTION
**Pull request by aMFeriuSS**
**- What I did**
Pulse fetching from light or heave for request
**- How I did it**
Add new cache for pulse-data into logicrunner. 
Add some new handlers and procedures on light and heavy for pulse-fetching

### CODE REVIEW

File: ledger/light/handle/get_pulse_test.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+	err := handler.Present(ctx, flow.NewFlowMock(t))
+	require.Error(t, err)
+}
+
+func TestGetPulse_IncorrectTypeMsgPayload(t *testing.T) {
+	t.Parallel()
+
+	ctx := inslogger.TestContext(t)
+	f := flow.NewFlowMock(t)
```
Comment (SergeyRadist) [inline ledger/light/handle/get_pulse_test.go]: better change to mc

File: ledger/light/handle/get_pulse_test.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+	ctx := inslogger.TestContext(t)
+	meta := payload.Meta{
+		Polymorph: uint32(payload.TypeMeta),
+		Payload:   nil,
+	}
+
+	handler := handle.NewGetPulse(nil, meta)
+
+	err := handler.Present(ctx, flow.NewFlowMock(t))
```
Comment (SergeyRadist) [inline ledger/light/handle/get_pulse_test.go]: better change to mc

File: ledger/light/proc/get_pulse_test.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+		coordinator.HeavyMock.Return(&insolar.Reference{}, nil)
+		reps := make(chan *message.Message, 1)
+		reps <- payload.MustNewMessage(&payload.Meta{
+			Polymorph: uint32(payload.TypeMeta),
+			Payload: payload.MustMarshal(&payload.Error{
+				Polymorph: uint32(payload.TypeError),
+				Code:      payload.CodeNotFound,
+			}),
+		})
+		sender.
```
Comment (SergeyRadist) [inline ledger/light/proc/get_pulse_test.go]: O_o

File: ledger/light/handle/get_pulse_test.go (hunk 1 above)
Comment (aMFeriuSS) [reply ledger/light/handle/get_pulse_test.go]: Done

File: ledger/light/handle/get_pulse_test.go (hunk 2 above)
Comment (aMFeriuSS) [reply ledger/light/handle/get_pulse_test.go]: Done

File: ledger/light/proc/get_pulse_test.go (hunk 3 above)
Comment (aMFeriuSS) [reply ledger/light/proc/get_pulse_test.go]: Corrected

File: insolar/payload/payload.proto [hunk 4]
```diff
@@ -372,3 +372,15 @@ message UpdateJet {
...
+message GetPulse {
+    uint32 Polymorph = 16;
+
+    bytes PulseNumber = 20 [(gogoproto.customtype) = "github.com/insolar/insolar/insolar.PulseNumber", (gogoproto.nullable) = false];
+}
+
+message Pulse {
+    uint32 Polymorph = 16;
+
+    pulse.PulseProto Pulse = 24 [(gogoproto.nullable) = false];
```
Comment (ita-sammann) [inline insolar/payload/payload.proto]: Why pulse field number is 24 and not 20?

File: ledger/heavy/proc/send_pulse.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+func (p *SendPulse) Proceed(ctx context.Context) error {
+	getPulse := payload.GetPulse{}
+	err := getPulse.Unmarshal(p.meta.Payload)
+	if err != nil {
+		return errors.Wrap(err, "failed to unmarshal GetPulse message")
+	}
+
+	foundedPulse, err := p.dep.pulses.ForPulseNumber(ctx, getPulse.PulseNumber)
```
Comment (ita-sammann) [inline ledger/heavy/proc/send_pulse.go]: found?

File: insolar/payload/payload.proto (hunk 4 above)
Comment (aMFeriuSS) [reply insolar/payload/payload.proto]: changed

File: ledger/heavy/proc/send_pulse.go (hunk 5 above)
Comment (aMFeriuSS) [reply ledger/heavy/proc/send_pulse.go]: changed

File: ledger/heavy/proc/send_pulse_test.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+		pulses = pulse.NewAccessorMock(mc)
+	}
+
+	newProc := func(msg payload.Meta) *proc.SendPulse {
+		p := proc.NewSendPulse(msg)
+		p.Dep(pulses, sender)
+		return p
+	}
+
+	resetComponents()
```
Comment (andreyromancev) [inline ledger/heavy/proc/send_pulse_test.go]: We call it `setup` now and call it at the beginning of a sub-test.

File: ledger/heavy/proc/send_pulse_test.go [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+	t.Run("pulse not found", func(t *testing.T) {
+		p := newProc(payload.Meta{})
+		pulses.ForPulseNumberMock.Return(insolar.Pulse{}, pulse.ErrNotFound)
+
+		err := p.Proceed(ctx)
+		require.Error(t, err)
+		assert.Equal(t, pulse.ErrNotFound, errors.Cause(err))
+
+		mc.Finish()
```
Comment (andreyromancev) [inline ledger/heavy/proc/send_pulse_test.go]: Let's call it right after `setup` as `defer`.

File: ledger/light/handle/init.go [hunk 8]
```diff
@@ -136,6 +136,8 @@ func (s *Init) handle(ctx context.Context, f flow.Flow) error {
 		err = f.Handle(ctx, NewError(s.message).Present)
 	case payload.TypeHotObjects:
 		err = f.Handle(ctx, NewHotObjects(s.dep, meta).Present)
+	case payload.TypeGetPulse:
```
Comment (andreyromancev) [inline ledger/light/handle/init.go]: Also, should be callable in the past.

File: ledger/heavy/proc/send_pulse_test.go (hunk 6 above)
Comment (aMFeriuSS) [reply ledger/heavy/proc/send_pulse_test.go]: done

File: ledger/heavy/proc/send_pulse_test.go (hunk 7 above)
Comment (aMFeriuSS) [reply ledger/heavy/proc/send_pulse_test.go]: done

File: ledger/light/handle/init.go (hunk 8 above)
Comment (aMFeriuSS) [reply ledger/light/handle/init.go]: done

File: logicrunner/artifacts/client.go [hunk 9]
```diff
@@ -206,6 +206,43 @@ func (m *client) RegisterOutgoingRequest(ctx context.Context, request *record.Ou
 	return res, err
 }
 
+// GetPulseForRequest returns pulse data for pulse number from request.
+func (m *client) GetPulseForRequest(
```
Comment (ruz) [inline logicrunner/artifacts/client.go]: this method should not take request as argument. use pulse number.

File: logicrunner/artifacts/client.go [hunk 10]
```diff
@@ -206,6 +206,43 @@ func (m *client) RegisterOutgoingRequest(ctx context.Context, request *record.Ou
...
+	ctx context.Context, request insolar.Reference,
+) (PulseDescriptor, error) {
+	var err error
+	ctx, instrumenter := instrument(ctx, "GetPulseForRequest", &err)
+	defer instrumenter.end()
+
+	foundPulse, err := m.PulseAccessor.ForPulseNumber(ctx, request.GetLocal().Pulse())
+	if err != nil {
```
Comment (ruz) [inline logicrunner/artifacts/client.go]: I would rather reverse logic and get rid of quick paths quickly:
```
if err != nil && err != ErrNotFound {
  return ...err
}
if err == nil {
   return result
}
.... not found case ...
return final
```

File: configuration/logicrunner.go [hunk 11]
```diff
@@ -51,5 +53,6 @@ func NewLogicRunner() LogicRunner {
 			RunnerListen:   "127.0.0.1:7777",
 			RunnerProtocol: "tcp",
 		},
+		PulseLRUSize: 10,
```
Comment (ruz) [inline configuration/logicrunner.go]: i believe it's way too small. should be somewhere in between 100-1000

File: insolar/payload/payload.go [hunk 12]
```diff
@@ -431,6 +439,14 @@ func Unmarshal(data []byte) (Payload, error) {
...
 		err := pl.Unmarshal(data)
 		return &pl, err
+	case TypeGetPulse:
+		pl := GetPulse{}
+		err := pl.Unmarshal(data)
+		return &pl, err
+	case TypePulse:
+		pl := Pulse{}
+		err := pl.Unmarshal(data)
+		return &pl, err
```
Comment (ruz) [inline insolar/payload/payload.go]: you injected these two right in the middle of group "GetRequest", "GetRequestInfo" and "TypeRequest"

File: insolar/vm.go [hunk 13]
```diff
@@ -99,7 +99,39 @@ type LogicCallContext struct {
 	Caller          *Reference // Contract that made the call
 	CallerPrototype *Reference // Prototype (base class) of the caller
 
-	TraceID string // trace mark for Jaegar and friends
+	TraceID      string // trace mark for Jaeger and friends
+	pulseFetcher func(ctx context.Context, pn PulseNumber) (Pulse, error)
```
Comment (ruz) [inline insolar/vm.go]: function pointer IS NOT acceptable in LogicCallContext, this struct is transferred over network, we already have means to implement such helpers.

**Comment by ita-sammann** [comment]
Tests passed successfully but not reported to github: https://teamcity.insolar.io/viewLog.html?buildId=184481
Force merging.

**Pull request merged**